ocb3/src/lib.rs

```diff
@@ -0,0 +1,602 @@
+            unimplemented!()
+        }
+
+        // First, try to process many blocks at once.
+        let (processed_bytes, mut Offset_i, mut Checksum_i) = self.wide_decrypt(nonce, buffer);
+
+        let mut i = (processed_bytes / 16) + 1;
+
+        // Then, process the remaining blocks.
+        for C_i in buffer[processed_bytes..].chunks_exact_mut(16) {
+            let C_i = Block::from_mut_slice(C_i);
+            // Offset_i = Offset_{i-1} xor L_{ntz(i)}
+            inplace_xor(&mut Offset_i, &self.L[ntz(i)]);
+            // P_i = Offset_i xor DECIPHER(K, C_i xor Offset_i)
+            let P_i = C_i;
+            inplace_xor(P_i, &Offset_i);
+            self.cipher.decrypt_block(P_i);
+            inplace_xor(P_i, &Offset_i);
+            // Checksum_i = Checksum_{i-1} xor P_i
+            inplace_xor(&mut Checksum_i, P_i);
+
+            i += 1;
+        }
+
+        // Process any partial blocks.
+        if (buffer.len() % 16) != 0 {
+            let processed_bytes = (i - 1) * 16;
+            let remaining_bytes = buffer.len() - processed_bytes;
+
+            // Offset_* = Offset_m xor L_*
+            inplace_xor(&mut Offset_i, &self.Lstar);
+            // Pad = ENCIPHER(K, Offset_*)
+            let mut Pad = Block::default();
+            inplace_xor(&mut Pad, &Offset_i);
+            self.cipher.encrypt_block(&mut Pad);
+            // P_* = C_* xor Pad[1..bitlen(C_*)]
+            let C_star = &mut buffer[processed_bytes..];
+            let Pad = &mut Pad[..C_star.len()];
+            for (aa, bb) in C_star.iter_mut().zip(Pad) {
+                *aa ^= *bb;
+            }
+            // Checksum_* = Checksum_m xor (P_* || 1 || zeros(127-bitlen(P_*)))
+            let Checksum_rhs = &mut [0u8; 16];
+            Checksum_rhs[..remaining_bytes].copy_from_slice(&buffer[processed_bytes..]);
+            Checksum_rhs[remaining_bytes] = 0b1000_0000;
+            inplace_xor(&mut Checksum_i, Block::from_slice(Checksum_rhs));
+        }
+
+        self.compute_tag(associated_data, &mut Checksum_i, &Offset_i)
+    }
+
+    /// Encrypts plaintext in groups of WIDTH.
+    ///
+    /// Adapted from https://www.cs.ucdavis.edu/~rogaway/ocb/news/code/ocb.c
+    fn wide_encrypt(&self, nonce: &Nonce<NonceSize>, buffer: &mut [u8]) -> (usize, Block, Block) {
+        #[cfg(not(target_feature = "avx512vaes"))]
+        const WIDTH: usize = 2;
+        #[cfg(not(target_feature = "avx512vaes"))]
+        let split_into_blocks = crate::util::split_into_two_blocks;
+
+        #[cfg(target_feature = "avx512vaes")]
+        const WIDTH: usize = 4;
+        #[cfg(target_feature = "avx512vaes")]
+        let split_into_blocks = crate::util::split_into_four_blocks;
+
+        let mut i = 1;
+
+        let mut Offset_i = [Block::default(); WIDTH];
+        Offset_i[Offset_i.len() - 1] = initial_offset(&self.cipher, nonce, TagSize::to_u32());
+        let mut Checksum_i = Block::default();
+        #[allow(unsafe_code)]
+        unsafe {
```

Comment:
I don't think it makes much sense to add `#![deny(unsafe_code)]` and then `#[allow(unsafe_code)]` it for all `unsafe` blocks. Seems... kinda pointless, no? At worst, it can be seen as false advertisement: "look! our code does not have any unsafe! well, except here and hare".